x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/categorization/GrokPatternCreator.java

```diff
@@ -0,0 +1,244 @@
+     * a grok pattern that will match the category and pull out any likely fields.  The extracted
+     * fields are given pretty generic names, but unique within the grok pattern provided.  The
+     * expectation is that a user will adjust the extracted field names based on their domain
+     * knowledge.
+     */
+    public static String findBestGrokMatchFromExamples(String regex, Collection<String> examples) {
+
+        // The first string in this array will end up being the empty string, and it doesn't correspond
+        // to an "in between" bit.  Although it could be removed for "neatness", it actually makes the
+        // loops below slightly neater if it's left in.
+        //
+        // E.g., ".*?cat.+?sat.+?mat.*" -> [ "", "cat", "sat", "mat" ]
+        String[] fixedRegexBits = regex.split("\\.[*+]\\??");
+
+        // Create a pattern that will capture the bits in between the fixed parts of the regex
+        //
+        // E.g., ".*?cat.+?sat.+?mat.*" -> Pattern (.*?)cat(.+?)sat(.+?)mat(.*)
+        Pattern exampleProcessor = Pattern.compile(regex.replaceAll("(\\.[*+]\\??)", "($1)"), Pattern.DOTALL);
+
+        List<Collection<String>> inBetweenBits = new ArrayList<>(fixedRegexBits.length);
+        for (String example : examples) {
+            Matcher matcher = exampleProcessor.matcher(example);
+            if (matcher.matches()) {
+                assert matcher.groupCount() == fixedRegexBits.length;
+                // E.g., if the input regex was ".*?cat.+?sat.+?mat.*" then the example
+                // "the cat sat on the mat" will result in "the ", " ", " on the ", and ""
+                // being added to the 4 "in between" collections in that order
+                for (int groupNum = 1; groupNum <= matcher.groupCount(); ++groupNum) {
+                    if (inBetweenBits.size() < groupNum) {
+                        inBetweenBits.add(new ArrayList<>(examples.size()));
+                    }
+                    inBetweenBits.get(groupNum - 1).add(matcher.group(groupNum));
+                }
+            } else {
+                // We should never get here.  If we do it implies a bug in the original categorization,
+                // as it's produced a regex that doesn't match the examples.
+                assert matcher.matches() : exampleProcessor.pattern() + " did not match " + example;
+            }
+        }
+
+        Map<String, Integer> fieldNameCountStore = new HashMap<>();
+        StringBuilder overallGrokPatternBuilder = new StringBuilder();
+        // Finally, for each collection of "in between" bits we look for the best Grok pattern and incorporate
+        // it into the overall Grok pattern that will match the each example in its entirety
+        for (int inBetweenBitNum = 0; inBetweenBitNum < inBetweenBits.size(); ++inBetweenBitNum) {
+            // Remember (from the first comment in this method) that the first element in this array is
```

Comment:
My understanding is that this causes every generated grok pattern to start with '.*?' unless one of `ORDERED_CANDIDATE_GROK_PATTERNS ` matches the empty string is that the intention?

Reply:
Because the Grok patterns are built from the regex that categorization churns out there's no way to be sure what comes before the first common token for the category.  Therefore all the regexes will begin with `.*?`.  (It doesn't make any difference if one of the `ORDERED_CANDIDATE_GROK_PATTERNS` matches the empty string - this prefix will _always_ be on the regex.)

I agree it could be inefficient and look silly to a human who knows the data, but this is a limitation from trying to reverse engineer Grok patterns from the categorization output.  An alternative would be to try to find fields as part of the main categorization algorithm.  But that would imply either moving the whole categorization algorithm into Java, or adding Grok functionality to the C++, or using something other than Grok to define how to extract the fields.

Possibly the best thing to do is just release this functionality in its current form as experimental, because several people have said they could get value from this prototype, and switch to something completely different and more complex in the future.